tools/dev/msftidy_docs.rb

```diff
@@ -0,0 +1,287 @@
+  end
+
+  #
+  # Display an info message. Info messages do not alter the exit status.
+  #
+  def info(txt, line=0)
+    return if SUPPRESS_INFO_MESSAGES
+    line_msg = (line>0) ? ":#{line}" : ''
+    puts "#{@full_filepath}#{line_msg} - [#{'INFO'.cyan}] #{cleanup_text(txt)}"
+  end
+
+  ##
+  #
+  # The functions below are actually the ones checking the source code
+  #
+  ##
+
+  def has_module
+    module_filepath = @full_filepath.sub('documentation/','').sub('/exploit/', '/exploits/')
+    found = false
+    ['.rb', '.py', '.go'].each do |ext|
+      if File.file? module_filepath.sub(/.md$/, ext)
+        found = true
+        break
+      end
+    end
+    unless found
+      error("Doc missing module.  Check file name and path(s) are correct. Doc: #{@full_filepath}")
+    end
+  end
+
+  def check_start_with_vuln_app
+    unless @lines.first =~ /^## Vulnerable Application$/
+      warn('Docs should start with ## Vulnerable Application')
+    end 
+  end
+
+  def has_h2_headings
+    has_vulnerable_application = false
+    has_verification_steps = false
+    has_scenarios = false
+    has_options = false
+    has_bad_description = false
+    has_bad_intro = false
+
+    @lines.each do |line|
+      if line =~ /^## Vulnerable Application$/
+        has_vulnerable_application = true
+        next
+      end
+
+      if line =~ /^## Verification Steps$/
+        has_verification_steps = true
+        next
+      end
+
+      if line =~ /^## Scenarios$/
+        has_scenarios = true
+        next
+      end
+
+      if line =~ /^## Options$/
+        has_options = true
+        next
+      end
+
+      if line =~ /^## Description$/
+        has_bad_description = true
+        next
+      end
+
+      if line =~ /^## (Intro|Introduction)$/
+        has_bad_intro = true
+        next
+      end
+    end
+
+    unless has_vulnerable_application
+      warn('Missing Section: ## Vulnerable Application')
+    end
+
+    unless has_verification_steps
+      warn('Missing Section: ## Verification Steps')
+    end
+
+    unless has_scenarios
+      warn('Missing Section: ## Scenarios')
+    end
+
+    unless has_options
+      warn('Missing Section: ## Options')
+    end
+
+    unless has_bad_description
+      warn('Descriptions should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application')
+    end
+
+    unless has_bad_intro
+      warn('Intro/Introduction should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application')
+    end
+  end
+
+  def check_newline_eof
+    if @source !~ /(?:\r\n|\n)\z/m
+      warn('Please add a newline at the end of the file')
+    end
+  end
+
+  # This checks that the H2 headings are in teh right order.
+  def h2_order
+    unless @source =~ /^## Vulnerable Application$.+^## Verification Steps$.+^## Options$.+^## Scenarios$/m
+      warn('H2 headings in incorrect order.  Should be: Vulnerable Application, Verification Steps, Options, Scenarios')
```

Comment:
This is not aware of missing sections.

Reply:
correct.  #6 I asked for @tperry-r7  thoughts on if its OK to have a missing section.  I think Options is most likely the only section that is commonly missing, and I'm personally ok with it as most modules don't have an option that's outside of the 'ordinary'.

Reply:
the other issue is half the comments about things are in PRs the other half in slack,s o also getting lost on where items are getting mentioned or not.

Reply:
I took both warnings into account, so I think it's okay for now. I understood what was intended.

Reply:
Fixed in #15140.